File: course_101/exercises/easy_7/staggered_caps_part_2.rb
# Modify the method from the previous exercise so it ignores non-alphabetic characters when determining whether it should uppercase or lowercase each letter.
# The non-alphabetic characters should still be included in the return value; they just don't count when toggling the desired case.

def staggered_case(string)
  result = ''
  need_upper = true
  string.chars.each do |char|
    if char =~ /[a-zA-Z]/
      if need_upper
        result << char.upcase
      else
        result << char.downcase
      end
      need_upper = !need_upper  
    else
      result << char
    end
  end
  result
end

# Further Exploration
#Modify this method so the caller can determine whether non-alphabetic characters should be counted when determining the upper/lowercase state.
# That is, you want a method that can perform the same actions that this method does,
# or operates like the previous version.

def staggered_case(string, ignore_non_alphabetic = true)
  result = ''
  need_upper = true
  string.chars.each do |char|
    if char =~ /[a-zA-Z]/ || !ignore_non_alphabetic
      if need_upper
        result << char.upcase
      else
        result << char.downcase
      end
      need_upper = !need_upper  
    else
      result << char
    end
  end
  result
end


puts staggered_case('I Love Launch School!') == 'I lOvE lAuNcH sChOoL!'
puts staggered_case('ALL CAPS') == 'AlL cApS'
puts staggered_case('ignore 77 the 444 numbers') == 'IgNoRe 77 ThE 444 nUmBeRs'

File: course_js210_computational_thinking_and_javascript_programming/exercises/js210_small_problems/string_and_text_processing/staggered_caps_part_2.js
// Modify the function from the previous exercise so that it ignores non-alphabetic characters when determining whether a letter should be upper or lower case. Non-alphabetic characters should still be included in the output string, but should not be counted when determining the appropriate case.

function staggeredCase(string) {
  let needCaps = true;
  let newChar;
  
  return string.split('').map(function (char) {
    if (/[a-z]/i.test(char)) {
      if (needCaps) {
        newChar = char.toUpperCase();
      } else {
        newChar = char.toLowerCase();
      }

      needCaps = !needCaps;
      return newChar;
    } else {
      return char
    }
  }).join('');
}

console.log(staggeredCase('I Love Launch School!'));        // "I lOvE lAuNcH sChOoL!"
console.log(staggeredCase('ALL CAPS'));                     // "AlL cApS"
console.log(staggeredCase('ignore 77 the 444 numbers'));    // "IgNoRe 77 ThE 444 nUmBeRs"